### PR TITLE
Add and use ConfigureAwaitOptions

### PR DESCRIPTION
**Pull request by stephentoub**
Adds the ability to further control how awaits are performed, with the ConfigureAwaitOptions enum.  For .NET 8 this is just for `Task` / `Task<TResult>`, and for the latter, the `SuppressThrowing` option isn't supported (as it can make the TResult erroneous).

Mostly everything here works as I'd like.  The one thing I'm not happy about is this adds an extra branch to the configured awaiter's IsCompleted property that's now incurred on every `await task.ConfigureAwait(false)`.  That branch should be elidable by the JIT, as the input is invariably constant (`false`) and it should be able to then see that constant value through to the IsCompleted call immediately after the awaiter is constructed.  However, because in the async method the awaiter is later passed by ref to the method builder, it's considered address exposed, the struct won't be promoted, and the JIT loses track of the known constant value, such that even though everything is inlined, it still does the `(options & ForceYielding) != 0` check in IsCompleted.  I'm talking with @EgorBo about whether there are any options here.  It's not worth introducing a whole new set of awaitables/awaiters for this, but hopefully we can come up with a way in the existing ones to remove the few additional instructions incurred.
```diff
+xor      edx, edx
+mov      dword ptr [rbp-08H], edx
+test     byte  ptr [rbp-08H], 4
+jne      SHORT G_M000_IG04
 mov      rdx, gword ptr [rbp-10H]
 test     dword ptr [rdx+34H], 0x1600000
 jne      SHORT G_M000_IG07
```

On the positive front, I've used the new overload throughout the repo to replace various custom awaiters, use of Task.Run, and places we were catching all exceptions to suppress them with awaits.  Some of these just help to clean up the code; others have / enable meaningful perf improvements.  For example:

|         Method |         Toolchain | Cancelable |     Mean | Ratio | Allocated | Alloc Ratio |
|--------------- |------------------ |----------- |---------:|------:|----------:|------------:|
| ReadWriteAsync | \main\corerun.exe |      False | 3.856 us |  1.00 |     181 B |       1.000 |
| ReadWriteAsync |   \pr\corerun.exe |      False | 3.060 us |  0.78 |       1 B |       0.006 |
|                |                   |            |          |       |           |             |
| ReadWriteAsync | \main\corerun.exe |       True | 4.108 us |  1.00 |     184 B |        1.00 |
| ReadWriteAsync |   \pr\corerun.exe |       True | 3.415 us |  0.83 |      82 B |        0.45 |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.IO.Pipes;
using System.Threading.Tasks;
using System.Threading;

[MemoryDiagnoser(false)]
public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

    private readonly CancellationTokenSource _cts = new CancellationTokenSource();
    private readonly byte[] _buffer = new byte[1];
    private AnonymousPipeServerStream _server;
    private AnonymousPipeClientStream _client;

    [Params(false, true)]
    public bool Cancelable { get; set; }

    [GlobalSetup]
    public void Setup()
    {
        _server = new AnonymousPipeServerStream(PipeDirection.Out);
        _client = new AnonymousPipeClientStream(PipeDirection.In, _server.ClientSafePipeHandle);
    }

    [GlobalCleanup]
    public void Cleanup()
    {
        _server.Dispose();
        _client.Dispose();
    }

    [Benchmark(OperationsPerInvoke = 1000)]
    public async Task ReadWriteAsync()
    {
        CancellationToken ct = Cancelable ? _cts.Token : default;
        for (int i = 0; i < 1000; i++)
        {
            ValueTask<int> read = _client.ReadAsync(_buffer, ct);
            await _server.WriteAsync(_buffer, ct);
            await read;
        }
    }
}
```

### CODE REVIEW

File: src/libraries/System.Threading.Tasks/tests/System.Runtime.CompilerServices/TaskAwaiterTests.cs [hunk 1]
```diff
@@ -561,6 +561,56 @@ public static IEnumerable<object[]> CanceledTasksAndExpectedCancellationExceptio
...
+        {
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("options", () => Task.CompletedTask.ConfigureAwait(options));
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("options", () => Task.FromResult(true).ConfigureAwait(options));
+        }
+
+        [Fact]
+        public void ConfigureAwaitOptions_SuppressThrowingUnsupportedOnGenericTask()
+        {
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("options", () => Task.FromResult(true).ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing));
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("options", () => Task.FromResult(true).ConfigureAwait((ConfigureAwaitOptions)0x7));
```
Comment (eiriktsarpalis) [inline src/libraries/System.Threading.Tasks/tests/System.Runtime.CompilerServices/TaskAwaiterTests.cs:576]: Don't recall the actual values of the enum, but presumably `0x7` contains the bits of `SuppressThrowing`?

File: src/libraries/System.Threading.Tasks/tests/System.Runtime.CompilerServices/TaskAwaiterTests.cs [hunk 2]
```diff
@@ -561,6 +561,56 @@ public static IEnumerable<object[]> CanceledTasksAndExpectedCancellationExceptio
...
+
+            if ((options & ConfigureAwaitOptions.SuppressThrowing) == 0)
+            {
+                Assert.False(new TaskCompletionSource<string>().Task.ConfigureAwait(options).GetAwaiter().IsCompleted);
+                Assert.False(Task.FromResult(42).ConfigureAwait(options).GetAwaiter().IsCompleted);
+            }
+        }
+
+        [Fact]
+        public void ConfigureAwaitOptions_SuppressThrowing_NoExceptionsAreThrown()
```
Comment (eiriktsarpalis) [inline src/libraries/System.Threading.Tasks/tests/System.Runtime.CompilerServices/TaskAwaiterTests.cs:597]: Could you also add a test for `ContinueOnCapturedContext` ?

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/Common/src/System/Threading/AsyncOverSyncWithIoCancellation.cs [hunk 3]
```diff
@@ -65,27 +51,24 @@ public void UnsafeOnCompleted(Action continuation)
...
         {
-            // Create the work item state object.  This is used to pass around state through various APIs,
-            // while also serving double duty as the work item used to queue the operation to the thread pool.
-            var workItem = new AsyncOverSyncWithIoCancellation();
-
-            // Queue the work to the thread pool.  This is implemented as a custom awaiter that queues the
-            // awaiter itself to the thread pool.
-            await workItem;
+            // Queue the work to complete asynchronously.
+            await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);
```
Comment (campersau) [inline src/libraries/Common/src/System/Threading/AsyncOverSyncWithIoCancellation.cs:55]: Is there a difference between this and `await Task.Yield().ConfigureAwait(false)` / `await Task.Yield().ConfigureAwait(ConfigureAwaitOptions.None)` ?

**Comment by HypsyNZ** [comment]
This is unacceptable behavior.

**Comment by HypsyNZ** [comment]
This is basically deliberately sabotage.

**Comment by HypsyNZ** [comment]
This literally sabotage and I'm reporting you to the authorities.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs [hunk 4]
```diff
@@ -88,6 +88,8 @@ internal void CancelOperation()
         // This is the async callback function. Only one thread could/should call this.
         private async Task ReadBufferAsync()
         {
+            await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);
```
Comment (davidfowl) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs:91]: Is this Task.Yield but skipping the sync context?

**Comment by HypsyNZ** [comment]
You obviously just did this to slow down earlier versions of .NET that are superior to the garbage you are pushing on everyone.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs:91]: Yes

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 5]
```diff
@@ -773,12 +776,7 @@ private void CheckForHttp2ConnectionInjection()
                 _pendingHttp2Connection = true;
 
                 RequestQueue<Http2Connection?>.QueueItem queueItem = _http2RequestQueue.PeekNextRequestForConnectionAttempt();
-
-                // Queue the creation of the connection to escape the held lock
-                ThreadPool.QueueUserWorkItem(static state =>
-                {
-                    _ = state.thisRef.AddHttp2ConnectionAsync(state.queueItem); // ignore returned task
-                }, (thisRef: this, queueItem), preferLocal: true);
```
Comment (HypsyNZ) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: Does this mean the work is more likely to happen on the global queue?

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: No.

File: src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpResponseStream.Managed.cs [hunk 6]
```diff
@@ -159,7 +159,7 @@ internal void InternalWrite(byte[] buffer, int offset, int count)
 
         private async Task InternalWriteIgnoreErrorsAsync(byte[] buffer, int offset, int count)
         {
-            try { await _stream.WriteAsync(buffer.AsMemory(offset, count)).ConfigureAwait(false); }
+            try { await _stream.WriteAsync(buffer.AsMemory(offset, count)).ConfigureAwait(bool); }
```
Comment (stephentoub) [inline src/libraries/System.Net.HttpListener/src/System/Net/Managed/HttpResponseStream.Managed.cs:162]:
```suggestion
            try { await _stream.WriteAsync(buffer.AsMemory(offset, count)).ConfigureAwait(false); }
```

File: src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs [hunk 7]
```diff
@@ -60,8 +59,12 @@ internal static partial class X509ResourceClient
...
             try
             {
-                ret = await s_downloadBytes(uri, cts?.Token ?? default, async).ConfigureAwait(false);
-                return ret;
+                Task<byte[]?> task = s_downloadBytes(uri, cts?.Token ?? default, async);
+                await ((Task)task).ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing);
+                if (task.IsCompletedSuccessfully)
+                {
+                    return task.Result;
+                }
```
Comment (Sergio0694) [inline src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: Just trying to understand this more: isn't this semantically different than the previous code? Specifically: the previous code would catch the exception if the task failed, which caused it to be marked as observed. The new code instead would just never observe the exception if the task failed, as it's just returning the result from the task if it did complete successfully. Wouldn't this make it so that if the task fails, the exception wrapper would not get its finalizer suppressed, and that exception would end up being signaled from the `TaskScheduler.UnobservedTaskException` event, which wasn't the case before this change? 🤔

Is that just by design, or am I missing something that makes this not the case here?
Thank you! 🙂

Comment (stephentoub) [reply src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: https://github.com/dotnet/runtime/blob/da1da02bbd2cb54490b7fc22f43ec32f5f302615/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs#L121

Comment (Sergio0694) [reply src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: Ooh I had missed that! Will definitely have to keep in mind that for cases where you want to await without throwing _but also_ flow any exceptions to `TaskScheduler.UnobservedTaskException` (eg. we're doing this in the MVVM Toolkit), this new configurable task awaiter shouldn't be used then. Thank you for the additional info! 😄

Comment (stephentoub) [reply src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: In what situation would you want that? From my perspective, the event is there for cases where the dev simply didn't pay attention. Any gesture that says "I know what I'm doing" supresses it. Just because someone accesses the Exception property or allows the exception to be thrown doesn't mean they've actually reacted to the error... they could have done `_ = t.Exception;` or `catch {}`. I see `SuppressThrowing` as a similar gesture.

Comment (Sergio0694) [reply src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: It's used under one of the configurable options for the `[RelayCommand]` generator ([see docs here](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/relaycommand)). Consider:

```csharp
partial class MyViewModel
{
    [RelayCommand]
    private async Task DoStuffAsync()
    {
    }
}
```

This will cause the generated command to directly await the method. If it throws, it throws. This is the default behavior and matches the usual developer expectation. If you don't handle exceptions, your app goes down. In most cases, you'd just use this, manually add a `try/catch`, and do all the exception handling/logging yourself from there.

```csharp
partial class MyViewModel
{
    [RelayCommand(FlowExceptionsToTaskScheduler)]
    private async Task DoStuffAsync()
    {
    }
}
```

This will instead await the method by not throwing exceptions, which will instead just flow to `TaskScheduler.UnobservedTaskException`, but without crashing the app. This can be useful in cases where you eg. have some centralized logging system for exceptions from your entire app, from a handler to that event, and where you don't particularly care about a given command to fail (or at least, you don't want that to crash your app).

It's just to give developers a bit more flexibility. We've had several feedbacks from people using/liking this option 🙂

Comment (stephentoub) [reply src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: I'm surprised you want to use UnobservedTaskException for that. The infrastructure is obviously responsible for consuming the task in this case and should also be responsible for dealing with any failures and routing them appropriately. What do you do if the exception is thrown synchronously out of such a Task-returning method, or if it's not Task-returning at all, or if it's a ValueTask backed by an IValueTaskSource that won't trigger UnobservedTaskException? async void methods, for example, queue exceptions to be rethrown on the original SynchronizationContext so that the app model's unhandled exception event can treat those the same as anything else unhandled by an event handler.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ConfigureAwaitOptions.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    /// <summary>Options to control behavior when awaiting.</summary>
+    [Flags]
+    public enum ConfigureAwaitOptions
+    {
+        /// <summary>No options specified.</summary>
+        /// <remarks>
+        /// <see cref="Task.ConfigureAwait(ConfigureAwaitOptions)"/> with a <see cref="None"/> argument behaves
+        /// identically to using <see cref="Task.ConfigureAwait(bool)"/> with a <see langword="false"/> argument.
+        /// </remarks>
+        None = 0x0,
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ConfigureAwaitOptions.cs:15]: @gewarren, is this still the case? If the docs really can't handle remarks, I'd expect our tooling that seeds the docs from our XML comments to massage them appropriately, or for the doc rendering engine to do so, but not for them to just be ignored. cc: @carlossanlop

File: src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs (hunk 7 above)
Comment (Sergio0694) [reply src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: > "What do you do if the exception is thrown synchronously out of such a Task-returning method"

The command is just wrapping your async method, so the only other scenario where invoking that command might throw is if you throw from an event handler subscribed to one of the properties of the command that send a notification. But in that case a crash is expected, as it's the same exact behavior you'd get if any other of your handlers (say, for some property change elsewhere) was to just throw while running. It'd be weird if that _didn't_ crash the app.

> "if it's not `Task`-returning at all, or if it's a `ValueTask` backed by an `IValueTaskSource` that won't trigger `UnobservedTaskException`?"

You can only use `[RelayCommand] for methods that return a `Task` (or `void`). If you tried to use it over a method that returned something else (like a `ValueTask`), the generator would just issue a diagnostic producing an error.

> "`async void` methods"

If a user had an `async void` method with `[RelayCommand]`, that's just an error on their side, and if that method threw an exception there would be nothing we could do. But then again, this would be exactly the same behavior you'd get for any other event handler being `async void` in your app. One of the points of `[RelayCommand]` supporting `Task`-returning methods is specifically so that it can correctly handle exceptions, as well as offering other useful functionality on top (eg. concurrency control, running checks, etc.). This actually makes me thing we might want to introduce a new analyzer to warn on `async void` methods with `[RelayCommand]`... I've seen a few of these in the wild 🤔

Comment (stephentoub) [reply src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: My point about "async void" was simply that it unifies with the exception handling mechanism of the app model, e.g. whether or not this event handler is marked "async", the same behavior results:
image
I'm surprised you're trying to create a different mechanism.  I'm obviously not your end user, but if I was, I'd expect unhandled exceptions to end up in the same place, raising Application.ThreadException in WinForms, raising Dispatcher.UnhandledException in WPF, etc.  UnobservedTaskException was not introduced to be a general exception handling choke point; it was introduced as a back stop for places where you inadvertently neglected to wait on a task.  Since you're raising this case as an example of where you'd need to be careful with SuppressThrowing, you're obviously awaiting the task, which means you're going out of your way to use UnobservedTaskException as such a general exception handling choke point.

Comment (Sergio0694) [reply src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: > "I'm surprised you're trying to create a different mechanism"

But we're not doing that. Let me clarify. The default mechanism, which is when you just have your `Task`-returning method for your command, will have the command await that task behind the scenes (in what is ultimately an `async void` handler, since it's invoked by the UI via `ICommand.Execute(object)`). If that method throws an exception that isn't handled, the exception will bubble up and result in the same as you're showing above. The main difference is that with your (wrapped) method being asynchronous, you get additional benefit on top, such as the command being able to also notify when an operation is running (which is very convenient to bind to eg. a loading ring, or to disable other commands), or to handle concurrency (eg. you can have the command automatically disable itself if it's already running). But with respect to exception handling, it's the same as what you're describing 🙂

The difference is only _if you opt-in_ to `FlowExceptionsToTaskScheduler`, in which case yes the behavior is different. But that's just a separate option, disabled by default, which you can just enable in specific cases if you do want that and know what you're doing and what it will result in. In almost all cases, you would just leave it off.

Comment (stephentoub) [reply src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: >> I'm surprised you're trying to create a different mechanism

> But we're not doing that. [...] The difference is only if you opt-in to FlowExceptionsToTaskScheduler, in which case yes the behavior is different

That's the "different mechanism" I'm referring to.  We never intended TaskScheduler.UnobservedTaskException to be used in this manner, as something folks would opt-in to sending lots of exceptions to.  If you're awaiting the task in your infra, and you don't want to allow the exception to bubble up to the app model's mechanism, I'd expect MVVM toolkit to have its own event handler folks can listen to for that use.

Anyway, thanks for the info.

Comment (HypsyNZ) [reply src/libraries/Common/src/System/Net/Http/X509ResourceClient.cs:67]: The amount of production software this is going to break is hilarious.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs (hunk 5 above)
Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: @HypsyNZ, you've been posting and deleting a bunch of comments that are dismissive or rude, eg
"The amount of production software this is going to break is hilarious." 
Please either stop or be more constructive in your feedback. Thank you.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: Oh you would rather dismissive and rude comments were public so you could complain about it publicly instead of taking responsibility for the fact that this sudden and unexpected change is going to break software?

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: Yes, I am aware you can read comments that were deleted.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: > Oh you would rather dismissive and rude comments were public so you could complain about it publicly instead of taking responsibility for the fact that this sudden and unexpected change is going to break software?

This addition has zero impact on existing behavior. What exactly is it going to break?

And we have a code of conduct covering this repo. I ask that you respect it.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: I'm pretty sure you are breaking your own code of conduct right now but whatever.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: > Yes, I am aware you can read comments that were deleted. This post is very professional.

I'm subscribed to the repo, and like anyone subscribed, I get notified of all comments, so my inbox has been going up with your unhelpful remarks and borderline obscene snipes. Please stop.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: Do you realize how much `ConfigureAwait(false)` gets called?

Not going to affect existing code, lmao.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: > Do you realize how much ConfigureAwait(false) gets called?

Yes, I do.  How exactly do you believe this PR negatively impacts the behavior of `ConfigureAwait(false)` such that it would break anyone?

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: You are just going to argue in bad faith no matter what I say. You already know the impact, You did it.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: And sorry to everyone else I didn't realize my comments got delivered to everyone geez that is embarassing.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: > You are just going to argue in bad faith no matter what I say. You already know the impact, You did it.

I honestly have no idea what you're talking about.  This PR adds a new overload that provides _additional_ functionality.  It does not change the existing functionality.  All code using `ConfigureAwait` today continues operating exactly as it did before this change.  Please elaborate on exactly what are your specific concerns for breaking.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: Also, You posted every single comment so why would you say that there are worse ones?

What you are doing right now breaks Microsofts code of conduct, not just your own.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: I did not want this interaction in a public setting, You are unprofessional and angry about being called out.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: Where does it say you are allowed to be deliberately confrontational?

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: What are you talking about which comment would that be?

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: Anyway, The point is, You already know what you have done, There is no point discussing it.

Comment (Bosch-Eli-Black) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: @HypsyNZ Just chill and go for a walk, man 🙂

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: > I did not want this interaction in a public setting,

If your don't want to discuss publicly, then don't post your comments publicly.

> You are unprofessional and angry about being called out.

I'm not angry about being called out. I'm legitimately asking what you believe the problem is. If there's actually a functional problem, I want to fix it.

What I'm upset about is your posting on multiple PRs today with inappropriate comments. Multiple people have pinged me about it, concerned about your behavior.

> Anyway, The point is, You already know what you have done, There is no point discussing it.

No, I don't. Please tell me so we can discuss it and fix it is there actually a problem.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: > @HypsyNZ Just chill and go for a walk, man 🙂

This was marked as resolved yesterday, this whole conversation was unprovoked.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: If you wanted to discuss my comments then this review is not the place you are breaking the code of conduct.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: > If you wanted to discuss my comments then this review is not the place

You left a bunch of your comments on this PR, so I'm responding to them on this PR.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: I have no doubt in my mind you know exactly what you have done, hence the comments that I thought only you could see.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: > I have no doubt in my mind you know exactly what you have done, hence the comments that I thought only you could see.

For the last time, I do not know what functional problem you're alluding to. Please share the specific concern.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: Regardless of my comments what you are doing in this review is an embarrassment.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: > Regardless of my comments what you are doing in this review is an embarrassment.

I'm responding to someone raising concerns and calling them out on verbal abuse throughout the repo today. I'm not embarrassed.

Comment (HypsyNZ) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:781]: You should be.

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Android.cs [hunk 9]
```diff
@@ -128,19 +128,24 @@ private static void StopLoop()
...
                 {
                     CheckIfNetworkChanged();
                 }
             }
             catch (OperationCanceledException)
             {
             }
+            finally
+            {
+                timer.Dispose();
```
Comment (pentp) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Android.cs:147]: Why is an explicit `try-finally` + `Dispose()` better here?

Comment (pentp) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Android.cs:147]: At the IL level both will have two separate `try` blocks. The size difference seems to come from Roslyn adding a null check before calling `Dispose` (which it could elide by doing some basic code analysis) and also adding some weird useless code at the beginning of the try block (no idea where that comes from):
```
// sequence point: hidden
IL_0086: ldloc.0
IL_0087: ldc.i4.1
IL_0088: pop
IL_0089: pop
// sequence point: hidden
IL_008a: nop
```

Comment (stephentoub) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Android.cs:147]: It is slightly smaller, but I changed it mainly because I was already editing this method, and I find fewer such constructs in the source easier to reason about.  A single try/catch/finally vs a using around a try/catch.  It's not a big deal either way.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ConfigureAwaitOptions.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (gewarren) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ConfigureAwaitOptions.cs:15]: Yes unfortunately remarks on enum fields are currently ignored. The internal tracking item is here: https://dev.azure.com/ceapex/Engineering/_workitems/edit/801799
cc @kexugit

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ConfigureAwaitOptions.cs:15]: @gewarren, thanks, what would you recommend in the interim?

**Comment by TonyValenti** [comment]
Hi @stephentoub -
I've been thinking about ```SuppressThrowing``` as it relates to ```Task<T>``` and I wanted to toss an idea out.
I haven't tested the new code so maybe my suggestion is already what you're doing.

TLDR
This should throw:
```
async Task<int> WorkAsync(){ throw new NotImpelementedExcepition(); }

var value = await WorkAsync().ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing);
```
because I'm trying to use the return value.

And this should not:
```
async Task<int> WorkAsync(){ throw new NotImpelementedExcepition(); }

await WorkAsync().ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing);
```
Because I do not actually try to retrieve the invalid result.

This will be useful because it does not force me to try/catch simply because a method I'm calling has a return value that I'm not using.

**Comment by TonyValenti** [comment]
My comment was that the throw should happen not on ConfigureAwait but when the .Result is accessed.

**Comment by stephentoub** [comment]
> My comment was that the throw should happen not on ConfigureAwait but when the .Result is accessed.

`await` expands into accessing `GetResult` on the awaiter, regardless of whether `GetResult` returns a value or that value is used.

**Comment by stephentoub** [comment]
> I don't see in [12.9.8.4 Run-time evaluation of await expressions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#12984-run-time-evaluation-of-await-expressions) any way for the awaiter to know whether the result of `await t` is going to be discarded. GetResult() will be called in either case. `await` does not read a Result property.

Correct

File: src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ConfigureAwaitOptions.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (gewarren) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ConfigureAwaitOptions.cs:15]: @stephentoub I think the two options are 1) Put the remarks in the summary for the field or 2) Put all the field remarks in the remarks for the enum itself. If option 1) is too heavyweight for IntelliSense, then I'd go with 2).

**Comment by Bosch-Eli-Black** [comment]
What would the below code print? Or would it throw?

```c#
async Task<string> WorkAsync() { throw new NotImpelementedException(); }

var value = await WorkAsync().ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing);

Console.WriteLine(value);
```

**Comment by stephentoub** [comment]
> What would the below code print? Or would it throw?

With:
```C#
var value = await WorkAsync().ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing);
```
the call to ConfigureAwait would throw a NotSupportedException.

With:
```C#
var value = await ((Task)WorkAsync()).ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing);
```
it wouldn't compile because the expression is void.

**Comment by Bosch-Eli-Black** [comment]
Thanks! 🙂
